### PR TITLE
fix is_wp_version_compatible() to accommodate incorrect format x.x.0

### PR DESCRIPTION
**Pull request by afragen**
Some users improperly test for releases of the format x.x.0. Unfortunately there is never a WordPress release with this version number and `version_compare( '5.2', '5.2.0', '>=' )` is false.

Test value sent to `is_wp_version_compatible( $required )` to ensure a trailing zero (x.x.0) is fixed to be (x.x) ~and a `wp_trigger_error()` is sent.~

This will correctly test the improper version x.x.0 as x.x.

Trac ticket: https://core.trac.wordpress.org/ticket/59448

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/functions.php [hunk 1]
```diff
@@ -8716,6 +8716,16 @@ function is_wp_version_compatible( $required ) {
 	// Strip off any -alpha, -RC, -beta, -src suffixes.
 	list( $version ) = explode( '-', $wp_version );
 
+	if ( is_string( $required ) && substr_count( $required, '.' ) > 1 && str_ends_with( $required, '.0' ) ) {
+		error_log(
```
Comment (aaronjorbin) [inline src/wp-includes/functions.php]: This should use `wp_trigger_error`

File: src/wp-includes/functions.php [hunk 2]
```diff
@@ -8716,6 +8716,16 @@ function is_wp_version_compatible( $required ) {
...
+			__FUNCTION__ . '(): ' .
+			/* translators: s: version string sent to function */
+			sprintf( __( '`%s` Not a valid WordPress version string.' ), $required )
```
Comment (aaronjorbin) [inline src/wp-includes/functions.php]: It would be helpful to present both what was passed in and what the result was. What do you think of this instead?

"Invalid version 6.0.0 passed to `is_wp_version_compatible`, 6.0 is assumed."

**Review by aaronjorbin — COMMENTED** [review]
Great start. I think this might need a test to ensure that the notice is properly raised. I also left a couple of comments on the specific implementation.

File: src/wp-includes/functions.php (hunk 1 above)
Comment (afragen) [reply src/wp-includes/functions.php]: I was initially having issues with tests not passing due to the `wp_trigger_error()` message being returned but found a solution by removing the error handler before the test and resetting afterwards.

**Comment by afragen** [comment]
Thanks @aaronjorbin, I believe I've fixed the issues you've raised. 

I don't quite understand what you mean by ensuring the notice is properly raised?

**Comment by aaronjorbin** [comment]
Thanks, the code changes look great.  Also, glad to hear this helped find another spot where our tests could be improved! 

> I don't quite understand what you mean by ensuring the notice is properly raised

I meant using something like what was implemented in https://core.trac.wordpress.org/changeset/57332 Essentially rather than silencing the error, setting an expect for it to be raised.

**Comment by afragen** [comment]
The problem with raising the error to an Exception is that the comparison will then fail and what we've intended is for the provided incorrect version be modified so it's now correct and the comparison continues -- or am I missing something?

**Comment by aaronjorbin** [comment]
Sorry I wasn't clear. I pushed some code to your branch since I thought showing the code would help.

By using expectException, the exception won't cause the tests to fail.  I did move the actual call to `is_wp_version_compatible` to it's own line so that the error handler can be restored to it's previous state.

Let me know if this makes sense. It looks like I'll need to clean up some formatting as well, so I can try to add some comments if you think it would help

File: tests/phpunit/tests/functions/isWpVersionCompatible.php [hunk 3]
```diff
@@ -12,14 +12,34 @@ class Tests_Functions_IsWpVersionCompatible extends WP_UnitTestCase {
...
 	 * @ticket 54257
 	 *
 	 * @param mixed $required The minimum required WordPress version.
 	 * @param bool  $expected The expected result.
 	 */
-	public function test_is_wp_version_compatible( $required, $expected ) {
-		$this->assertSame( $expected, is_wp_version_compatible( $required ) );
+	public function test_is_wp_version_compatible( $required, $expected, $expect_exception = false ) {
+
+		if ( $expect_exception ) {
```
Comment (costdev) [inline tests/phpunit/tests/functions/isWpVersionCompatible.php]: These latest changes provide a false sense of security, as does the test in [this Core changeset](https://core.trac.wordpress.org/changeset/57332).

For example, if you comment out all datasets _except_ those that should throw an exception, and change the final assertion to `$this->assertSame( $expected . 'should not pass', $result );`, the two datasets will still pass. This is because an exception was correctly detected, but the test also ended as soon as this occurred. This issue also exists with the test in the Core changeset linked above.

Instead, let's consider this approach:

1. Add a new property to the test class:

```php
/**
 * A notice expected during the tests.
 *
 * @var string
 */
private $expected_notice = '';
```
2. Add a `tear_down()` method that resets the property:
```php
/**
 * Resets the `expected_notice` property after each test runs.
 */
public function tear_down() {
	$this->expected_notice = '';
	parent::tear_down();
}
```
3. Adjust the test so that instead of throwing an exception, we note the notice, restore the previous handler, and check that it matches the expected type and string:
```php
public function test_is_wp_version_compatible( $required, $expected, $expect_notice = false ) {
	if ( $expect_notice ) {
		set_error_handler(
			function ( $errno, $errstr ) {
				$this->expected_notice = $errno . ' ' . $errstr;
			},
			E_ALL
		);
	}

	$result = is_wp_version_compatible( $required );

	if ( $expect_notice ) {
		restore_error_handler();
		$this->assertStringStartsWith(
			E_USER_NOTICE . " is_wp_version_compatible(): `$required` Not a valid WordPress version string.",
			$this->expected_notice,
			'The expected notice was not thrown.'
		);
	}

	$this->assertSame( $expected, $result );
}
```

Now, if we change the final assertion to `$this->assertSame( $expected . 'should not pass', $result );`, the two datasets will correctly fail.

Run a few checks locally to see the issue with the existing test, and how the changes above resolve this.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php [hunk 4]
```diff
@@ -12,14 +12,34 @@ class Tests_Functions_IsWpVersionCompatible extends WP_UnitTestCase {
 	 * Tests is_wp_version_compatible().
 	 *
 	 * @dataProvider data_is_wp_version_compatible
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/functions/isWpVersionCompatible.php]: Why are these needed?

File: src/wp-includes/functions.php [hunk 5]
```diff
@@ -8735,6 +8735,16 @@ function is_wp_version_compatible( $required ) {
...
 	list( $version ) = explode( '-', $wp_version );
 
+	if ( is_string( $required ) && substr_count( $required, '.' ) > 1 && str_ends_with( $required, '.0' ) ) {
+		$passed   = $required;
+		$required = preg_replace( '/\.0/', '', $passed, 1 );
+		wp_trigger_error(
+			__FUNCTION__,
+			/* translators: s: version string sent to function */
+			sprintf( __( '`%1$s` Not a valid WordPress version string. `%2$s` is assumed.' ), $passed, $required )
+		);
```
Comment (peterwilsoncc) [inline src/wp-includes/functions.php]: Per notes on ticket -- will need changes to the tests.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php [hunk 6]
```diff
@@ -85,6 +105,34 @@ public function data_is_wp_version_compatible() {
...
+				true,
+			),
+			'correct version x.0'       => array(
+				'required' => '5.0',
+				'expected' => true,
+			),
+			'correct version x.0.1'     => array(
+				'required' => '5.0.1',
+				'expected' => true,
+			),
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/functions/isWpVersionCompatible.php]: Is it worth adding a test to ensure the version is converted correctly? It might require some code shuffling.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 4 above)
Comment (costdev) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: These shouldn't be needed as there are no constants/static variables at play.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 6 above)
Comment (costdev) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: Provided there are tests covering all the bases, testing the output with each input should be enough, and the respective lines can be checked for coverage in a coverage report.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 3 above)
Comment (costdev) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: Looks like the discussion has moved away from having a notice. The above can be disregarded in that case. 🙂

File: src/wp-includes/functions.php (hunk 5 above)
Comment (afragen) [reply src/wp-includes/functions.php]: Are we planing on notifying the user in this manner or silently fixing the incorrect version?

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 6 above)
Comment (afragen) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: I added this test to ensure a `.0` in the middle of a version number wouldn’t be changed.

File: src/wp-includes/functions.php (hunk 5 above)
Comment (peterwilsoncc) [reply src/wp-includes/functions.php]: Silently fix due to the use of `@since x.x.0` throughout core. 

```suggestion
```

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 6 above)
Comment (afragen) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: I will add some new tests for that. I'll have it fixed by tomorrow.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 4 above)
Comment (afragen) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: Removed

File: tests/phpunit/tests/functions/isWpVersionCompatible.php (hunk 3 above)
Comment (costdev) [reply tests/phpunit/tests/functions/isWpVersionCompatible.php]: Un-resolving this as a decision on the notice hasn't finalized.

@aaronjorbin FYI: Take a look at the first comment in this thread regarding an issue with the changeset you referenced and previous commits to this PR that were influenced by it, and an alternative without this issue.

**Comment by azaozz** [comment]
Thinking this code is far too complex for what it needs to accomplish. We only care to strip `.0` from the end of `$required`, when it is longer than 3 characters, nothing more. So `substr_count()` and `preg_replace()` for a simple string that contains only digits and dots is perhaps "too much"? 

Perhaps something simpler would also work:
```php
if ( is_string( $required ) && strlen( $required ) > 3 && str_ends_with( $required, '.0' ) ) {
    // Strip trailing zero to properly match two digits (major) WP version numbers.
    $required = substr( $required, 0, -2 );
}
```

The `strlen( $required ) > 3` is there to catch versions like 5.0, 6.0, etc. In all other cases the WP versions should not have a trailing zero.

Edit: actually this should probably be `strlen( $required ) > 4` to catch WP versions up to 99.0 (that would still work properly). Seems reasonable to update this code when WP gets to 100.0 :)

**Review by azaozz — CHANGES_REQUESTED** [review]
See https://github.com/WordPress/wordpress-develop/pull/5677#issuecomment-1958086757.

**Comment by peterwilsoncc** [comment]
@azaozz I've switched `substr` to remove the version number's trailing `.0`

`$required` is developer input that can't be trusted to be entirely correct (the entire point of this ticket ;) so I've kept string count.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php [hunk 7]
```diff
@@ -8,6 +8,21 @@
...
+	private $expected_notice = '';
+
+	/**
+	 * Resets the `$expected_notice` property after each test runs.
+	 */
+	public function tear_down() {
+		$this->expected_notice = '';
+		parent::tear_down();
+	}
+
```
Comment (costdev) [inline tests/phpunit/tests/functions/isWpVersionCompatible.php]:
```suggestion
```

@peterwilsoncc you missed a spot 🧹

**Review by azaozz — APPROVED** [review]
> $required is developer input that can't be trusted to be entirely correct...

Right. Thinking it looks better now, and perhaps a very tiny bit faster (afaik PHP needs some time to "initialize" the external lib every time a regexp function is used) :) 
Seems this is good to go if there isn't anything else.

File: tests/phpunit/tests/functions/isWpVersionCompatible.php [hunk 8]
```diff
@@ -43,51 +43,140 @@ public function data_is_wp_version_compatible() {
...
-			'empty array'               => array(
+			'empty array'                     => array(
 				'required' => array(),
 				'expected' => true,
 			),
 		);
 	}
 
+	/**
+	 * Tests that is_wp_version_compatible() throws a notice for incorrect version numbering.
```
Comment (swissspidy) [inline tests/phpunit/tests/functions/isWpVersionCompatible.php]:
```suggestion
	 * Tests that is_wp_version_compatible() gracefully handles incorrect version numbering.
```

**Comment by peterwilsoncc** [comment]
Merged [r57707](https://core.trac.wordpress.org/changeset/57707) / https://github.com/WordPress/wordpress-develop/commit/2ba8f942cb8e95bf93d19c12f9cd9bb9ae61b72e
